hwangsae/relay.c

```diff
@@ -452,16 +452,30 @@ _relay_main (gpointer data)
 
               while (it) {
                 SRTSOCKET source_socket = GPOINTER_TO_INT (it->data);
+                gint sent = 0;
+                gssize len = 0;
+                gint rest = MIN (recv - len, 1316);
 
                 it = it->next;
+                while (len < recv) {
+                  if ((sent =
+                          srt_send (source_socket, (char *) (buf + len),
+                              rest)) < 0) {
```

Comment:
Sometimes, `sent` is less than `recv`, but in that case, `srt_send()` can cause an error in second-try.
In addition, repeating `srt_send()`  adds some delay to next `source_socket`.

Reply:
Is waiting for the full buffer to be sent at the expense of all other connected viewers the right course of action though? One viewer with less reliable connection means the relay may get stuck in this loop indefinitely while other clients starve.

When `srt_send()` sends less than is the message size it means libsrt's internal buffer is too small to keep up with the rate data is being written into the socket.

In my opinion seeing a lot of partially sent messages in the relay's log should be an indication for the administrator they should increase `SRTO_SNDBUF`. So we should rather make the send buffer size configurable so that it's size can be tuned according to the bandwidth requirements of a particular relay deployment.

Reply:
Thank you for good advice. In some network conditions, there are large packet drops (partial sending is just one of them). I am still investigating the root cause.

Reply:
There might be also a problem with the size of UDP socket send and/or receive buffer in the kernel that may lead to some data being silently dropped during transfer.